File: my_code/DSA/PYTHON/DS/stack and queue/Evaluate_Reverse_Polish_Notation_(Postfix).py
from math import ceil, floor

def evaluate_postfix(arr):
        stack = []

        for item in arr:
            # if the item is operator so pop the two numbers from the stack
            # then make the process and push the result instead of them
            if item == '+':
                stack.append(stack.pop() + stack.pop())
            elif item == '-':
                last, first = stack.pop(), stack.pop()
                stack.append(first - last)
            elif item == '*':
                stack.append(stack.pop() * stack.pop())
            elif item == '/':
                last, first = stack.pop(), stack.pop()
                stack.append(first // last)
            # if is a number push it to the stack
            else:
                stack.append(int(item))
        # because now we have just one value in the stack
        return stack[0]


print(evaluate_postfix(["2", "1", "+", "3", "*"]))

File: my_code/DSA/JS/DS/stack and queue/Evaluate_Reverse_Polish_Notation_(Postfix).js
function evaluate_postfix(arr) {
  const stack = [];

  for (const item of arr) {
    // If the item is an operator, pop two numbers from the stack,
    // perform the operation, then push the result back to the stack
    if (item === "+") {
      stack.push(stack.pop() + stack.pop());
    } else if (item === "-") {
      const last = stack.pop();
      const first = stack.pop();
      stack.push(first - last);
    } else if (item === "*") {
      stack.push(stack.pop() * stack.pop());
    } else if (item === "/") {
      const last = stack.pop();
      const first = stack.pop();
      // Use Math.floor to mimic integer division like Python's //
      stack.push(Math.floor(first / last));
    } else {
      // If it's a number, push it to the stack
      stack.push(parseInt(item));
    }
  }

  // After processing all items, the stack has one value — the result
  return stack[0];
}

console.log(evaluate_postfix(["2", "1", "+", "3", "*"])); // Output: 9
